Generic across forms (#5389)
carbon-language/carbon-lang: proposals/p5389.md

    and so on) of parameters or returns in the signature of functions.
-   A `form(`...`)` literal syntax for writing values of these form types.
-   Form facet types for expressing constraints on form types, which may be
    composed using `&`.
-   `:?` binding and `->?` return syntax parameterized by a form value.
-   Form versions of interfaces, facets, and facet types.
-   Use these new form facet types to create new "primitive" versions of the
    operator interfaces, like `Core.CallPrimitive`, that allow control over the
    form of the parameters and return.
-   A `call(`...`)` syntax for concisely and naturally writing facet types of
    the `Core.CallPrimitive` interface.
-   Impl lookup will allow some conversions between `Core.Form` values in
    interface arguments and associated constants, when specified by the
    interface, replacing the existing mechanism where `impl`s use `like` to
    allow implicit conversions in arguments.
-   A number of increased capabilities for named constraints.

Together these changes will allow:

-   the operator interfaces to support Carbon's full range of function
    semantics, particularly in the `Core.CallPrimitive` form interface;

> -   replacing our existing operator interfaces like `Core.Call` and
>     `Core.AddWidth` with named constraints that use types instead of forms;

Reviewer: This phrasing makes it sound like `Core.Call` etc. used to use forms. Maybe rephrase to something like this instead?

```suggestion
-   preserving the existing type-oriented constraints like `Core.Call`
    and `Core.AddWith`, now implemented as named constraints in
    terms of the form-oriented primitive interfaces, for common-case
    uses that don't need to deal with forms;
```

    conversions from the arguments values to the parameter types.
-   **Subsumption**: An overload set containing two functions, one accepting a
    `bool` and one accepting an `i32`, should be able to be passed as an
    argument to a function requiring a callable accepting a `bool`. This same
    mechanism should allow a generic function with signature that could accept a
    `bool`, such as `fn F[T:! type](x: T)`, to be passed to the function
    requiring a callable accepting a `bool`, using the instantiation with
    `T == bool`.
-   **Closed**: There is a way to say "just these implementations and no
    others." This should be the default when defining ordinary functions and
    function overloads.

However, fully addressing the last four of these properties is out of scope of
this proposal, since it won't address declaring overloaded functions and `impl`
prioritization. We still would like the decisions in this proposal to be
consistent with a future solution that achieves those properties.

In addition, there are a number of outstanding problems in Carbon that arise
from a construct requiring some specification of aspects in a binding when it

> should instead to a

Reviewer: Looks like something's missing here.

        runtime evaluation.
    -   `(`_form1_`, ` _form2_`, ` _form3_`) as Core.Form` creates a compound
        tuple form value.
    -   `{.a = `_form1_`, .b = ` _form2_`} as Core.Form` creates a compound
        struct form value.
-   that form values have a `.type` property that gives the unqualified (without
    `partial` or `const` modifiers) version of the type associated with the
    form.
-   form facet types that are restricted versions of `Core.Form`, such as
    `Core.NoValForm`.
-   a binding _name_`:?`_form value_, where the form of the binding is
    determined by the _form value_.
-   a return syntax `->? `_form value_, where the form of the return is
    determined by the _form value_.
-   form interfaces, where the `Self` is of type `Core.Form` or a form facet
    type instead of `type`.
-   that an `impl` of a form interface, or an interface with form parameters,
    uses the `.type` property of form arguments to determine the type to use for
    the orphan rule.
-   some uses of `extend` that previously assumed `Self` and didn't allow it to
    be specified, now allow `Self` or a form using `Self` to be specified.

> -   `Core.ValSelf`, `Cor.RefSelf`, and `Core.VarSelf` that take a `Core.Form`

Reviewer: ```suggestion
-   `Core.ValSelf`, `Core.RefSelf`, and `Core.VarSelf` functions that take a `Core.Form`
```

    facet type and return an ordinary `type` facet type with the specified
    category used for `Self`.

> -   `Core.Form where .type = T` which gives a form facet type that matches any

Reviewer: This seems like it follows from what's already been said about `Core.Form` and `.type`, rather than being a separate feature introduced by this proposal. Am I missing something?

    form with unqualified type `T`. Similarly, `Core.Form where .type impls C`
    gives a form facet type that matches any form whose type implements `C`.
-   `Core.ConvertsInto(PF:! Form)` which is a `Core.Form` constraint that
    matches all expression forms that can be converted to a parameter with form
    `PF`.
-   `implicit_into` which may be used to mark `Core.Form` interface parameters,
    including the `Self` parameter of a form interface. Those parameters are
    called "variant", specifically "contravariant." The arguments to a query in
    that position may be converted, as described by the `Core.ConvertsInto`
    constraint.
    -   An `impl` of such an interface can have multiple type structures,
        depending on whether the types in the query match or need to be
        converted.
    -   An `impl` may use `anchor` or `exact` to limit the conversions allowed,
        which can be needed to satisfy the orphan rule.
    -   Implied constraints are not added for variant parameters.

>     -   When normally we would for a facets of an interface, if that interface

Reviewer: I'm having a hard time parsing this sentence -- I feel like maybe there's a missing verb after "would", and "a facets" is clearly a typo but I'm not sure what the correct version is.

        has variant parameters, we create an adapter that also implements that

>         interface. This adapter performs the conversion between the argument an

Reviewer: ```suggestion
        interface. This adapter performs the conversion between the argument and
```

        parameter forms, and allows the implementations of the interface to
        differ from the original type without introducing coherence concerns.
-   `implicit_from` which similarly may be used to mark an interface parameter
    or associated constant. These parameters are also "variant," specifically
    "covariant." These allow conversions in the opposite direction as

>     `implicit_into`, as described by the `Core.ConvertsFrom` constraint.

Reviewer: This is the first mention of `Core.ConvertsFrom` -- was it supposed to have a bullet point like `Core.ConvertsInto`?

-   updates to named constraints to support forms.
    -   We add support for associated constants in named constraints. Their
        values must be deducible from an `impl` of an interface required by the
        named constraint.
    -   A named constraint can contain a `match_first` block to prioritize
        between requirements that overlap.
-   a change to the ...`(`...`)` function call operator. It now uses the
    `Core.CallPrimitive` form interface, instead of the `Core.Call` interface.
    -   A `call` sugar syntax makes it easier to specify a specific
        `Core.CallPrimitive` with a syntax similar to a function declaration
        instead of using form literals.
    -   The `Core.Call` interface is changed to a named constraint that uses

>         types to specify the call instead of forms.

Reviewer: Here again, the "instead of" sounds like it's contrasting with the old `Core.Call` rather than the current `Core.CallPrimitive`

    named constraints that allow you to continue to use types.
-   `auto` may now be used to deduce a form in a `:?` form binding or `->?` form
    return.
    -   This is used implicitly for positional parameters.
    -   This is used to give a new definition of the `alias` construct.
    -   This new `alias` construct doesn't apply to namespaces, so a new
        namespace aliasing facility is introduced.

## Proposal

### Form type and values

First, we introduce the new special type `Core.Form` that describe the form of
bindings (such as function parameters or in a `let` declaration), returns, or
expressions.

-   The new `:?` syntax takes a form value to
    [specify the form of a binding](#specifying-the-form-of-a-binding).
-   The new `->?` syntax takes a form value to
    [specify the form of a return](#specifying-the-form-of-a-return).

Values of `Core.Form` are compile-time only, like types. Form values can be
compound forms or primitive forms. A compound form is either a tuple form or
struct form.

A primitive `Core.Form` has:

-   a type
-   a category, which can be `val`/value, `ref`/reference, or `var`/initializing

> -   for parameter bindings, whether it is marked with `bound`

Reviewer: #5545 doesn't treat this as part of the form. Should I add it?

> -   the phase, which can be "compile time" or "run time" (meaning "when the

Reviewer: Are you intentionally excluding the symbolic vs. template/concrete distinction from the phase? If so, why?

    function is run")
-   for compile time phase, a specific compile-time value

#### Primitive form value literals

A primitive `Core.Form` value with "run time" phase can be written `form(`_cat_
_type_`)`, where _cat_ is an optional category modifier, and _type_ is a type
expression:

| `form` syntax | category           |
| :------------ | :----------------- |
| `form(val T)` | `val`/value        |
| `form(ref T)` | `ref`/reference    |
| `form(var T)` | `var`/initializing |

This was
[discussed on 2025-06-04](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.stp5dpvithkz).
We use `val` for the value category in alignment with the using `val` for value
category in returns in the resolution of
[leads issue #5522](https://github.com/carbon-language/carbon-lang/issues/5522).

#### `formof`

You may also write `formof(`_expression_`)` to get the form of the expression
argument. If the expression has "compile time" phase, then this will result in a

> form with that phase and also the value. If the argument to `formof` is a call

Reviewer: But what category? I ask because if it's the category of the expression, the definition of deduction in terms of `formof` seems to imply that `alias Two = 1 + 1;` is equivalent to `var Two:! auto = 1 + 1;`, and I'm having trouble working out if that's even legal, much less what the semantics would be.

```carbon
// Equivalent to: `fn Init() -> i32;`
fn Init() ->? form(var i32);

// Equivalent to: `fn Val() -> val i32;`
fn Val() ->? form(val i32);

// Equivalent to: `fn Ref() -> ref i32;`
fn Ref() ->? form(val i32);

// Deterimines the return form from the
// argument. `Param(form(ref i32))`
// returns an `i32` ref.
fn Param(RF:! Core.Form) ->? RF;
```

### Form interfaces

Previously, interfaces had an implicit `Self` type parameter, set to the type
that is implementing the interface. We add the capability to explicitly declare
the `Self` parameter with a form facet type as an implicit parameter in square
brackets `[`...`]`, similar to the `self` parameter of methods:

```carbon
interface Callback[Self:! Core.Form] {
  // Uses `Self` as the form of a
  // parameter using `:?`.
  fn Go[self:? Self]();
}

interface DefaultConstructible
    [Self:! Core.Form] {
  // Uses `Self` as the form of the
  // return using `->?`.
  fn MakeDefault() ->? Self;
}
```

`Callback` and `DefaultConstructible` are _form interfaces_, specifically
`Core.Form` interfaces in this case. A value of a form interface is a _form
facet_. These work the same way as ordinary "type" interfaces except they use

> the matching kind of form values instead of type values. An ordinary "type"

Reviewer: I'm having trouble following this sentence. Does "these" refer to form interfaces or form facets? What does "the matching kind of form values" mean, and in what sense does a form facet/interface use them?

> interface acts as if `Self` is declared `[Self:! type]` instead of with a form

Reviewer: Are you allowed to write that declaration explicitly? If so, must the type be `type` or can it be another non-form facet type?

type.

Let's break down what happens with an ordinary "type" interface:

```carbon
// Equivalent to: interface I[Self:! type]
interface I {
  fn F[self: Self]();
}

// Since `I` is effectively declared `I[Self:! type]`, that means
// `T` is a facet of a type, which means a type that is known to
// implement `I`. Since `T` is a facet of a type, it has an implicit
// conversion to `type`, so it can be used to the right of a `:`
// to give a type to `x`.
fn A[T:! I](x: T) {
  x.F():
  // - Looks in the type of `x` for a `F`

>   // - Type of `x` is `T` is symbolic, so look in its type instead

Reviewer: ```suggestion
  // - Type of `x` is `T`, which is symbolic, so look in its type instead
```

```

### Forms in `impl`s

Interfaces can also use form types as the types of its parameters or associated
constants.

```carbon
interface OneArg(P:! Core.Form) {
  let Result:! Core.Form;
  fn Fun[self: Self](a:? P) -> Result;
}

impl MyType as OneArg(form(val i32)) {
  where Result = form(var bool);
  fn Fun[self: Self](a: i32) -> bool;
}
```

To satisfy coherence by
[the orphan rule](/docs/design/generics/details.md#orphan-rule), forms are
converted to types in the type structure using the
[`.type` property](#type-property-of-form-values) (as
[discussed on 2025-06-11](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.rebuyhkgncpq)).
The form value itself is used, though, when prioritizing during `impl` lookup.

### `where`...`impls`

A `where` constraint may now contain a requirement `A impls B` where `A` is a
form or facet of a form, and `B` is a form facet type. For example,
`T:! type where form(val .Self) impls Callback` will bind to types `T` such that
there is an implementation of `Callback` for `form(val T)`.

### `extend`

Prior to this proposal, the `extend` modifier on some declarations means that

> some type argument may not be specified and instead is always set to `Self`:

Reviewer: It took me a while to understand that this section isn't proposing new syntaxes, just dropping a restriction on the syntaxes we already have.

```suggestion
some type argument must be omitted, and is presumed to be `Self`, although
there's a syntactic position available for it:
```

-   In an interface or named constraint, `extend require`, as defined in
    [proposal #5337](https://github.com/carbon-language/carbon-lang/pull/5337),
    must be followed by `impls` instead of allowing a type to be specified.
-   In a class, `extend impl` and `extend final impl` must be followed by `as`
    instead of allowing a type to be specified.
-   Similarly, in an interface, `extend impl` and `extend final impl`, as
    defined in
    [proposal #5337](https://github.com/carbon-language/carbon-lang/pull/5337),
    must be followed by `as` instead of allowing a type to be specified.

> In all of these cases, we add support for specifying either a type or a form.

Reviewer: ```suggestion
In all of these cases, we drop that restriction, and allow a type or a form to
be specified in the expected position.
```

Given a form facet type, we can construct an ordinary `type` facet type in a few
different ways. One is to construct a named constraint. The `Self` values of the
named constraint are types, but the requirements of the constraint can be on
form values using the `Self` type.

```carbon
// Equivalent to `constraint ValCallback[Self:! type]`
constraint ValCallback {
  extend require form(val Self) impls Callback;
}

constraint RefCallback {
  extend require form(ref Self) impls Callback;
}

constraint VarCallback {
  extend require form(var Self) impls Callback;
}
```

Another way is with [a `where`...`impls` expression](#whereimpls):

-   `type where form(val .Self) impls Callback`
-   `type where form(ref .Self) impls Callback`
-   `type where form(var .Self) impls Callback`

This however doesn't give a way to have name lookup into the facet type find
names in `Callback`, like the named constraints accomplish with
[`extend`](#extend). For that, we introduce functions that take a `Core.Form`
facet type and return an ordinary `type` facet type: `Core.ValSelf`,
`Core.RefSelf`, and `Core.VarSelf`. They are used like:

```carbon
fn InvokeValCallback

>     [T:! Core.ValSelf(Callback)](c: T) {

Reviewer: I find these names pretty opaque at the point of use. `ValSelfImpls(Callback)` might be better, but makes it even more tempting to write `type where form(val .Self) impls Callback` instead. Have we ever considered a core-language syntax for this, like `type where form(val .Self) extend impls Callback`?

from the `forall` clause. In this example:

```carbon
impl [PF:! Core.Form where .type = MyClass] PF as I;
```

The type structure for the orphan rule is `PF.type as I` which, after applying
the rewrite from `PF`'s constraint, becomes `MyClass as I`.

This approach also allows deduction of a type through a form, as in:

```carbon
fn F[T:! I](PF:! Core.Form where
            .type = Vector(T))
    -> T.Iterator;
```

since the `T` value is uniquely determined by `PF`.

This was
[discussed on 2025-06-11](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.rebuyhkgncpq),
along with alternative ways we could specify this constraint.

### `Core.ConvertsInto`

The `Core` package defines the `Form` facet type `ConvertsInto(PF:! Form)`,
which says which `Form`s describe a binding that would be allowed to convert to
the binding described by `PF`. So `form(val i32)` is a value of
`ConvertsInto(form(val i64))` since there is an implicit conversion from `i32`
to `i64`. More precisely, if we create a function `F` with `PF` as the form of
its parameter, and `y` a binding with form `SRC` (also a `Form` value), as in

```carbon
fn F(x:? PF);

> fn G(y:? SRC) {

Reviewer: Why not use a function `Y() ->? SRC` instead of a parameter `y:? SRC`? That seems simpler in several ways:
- forms "natively" describe expressions rather than bindings, so it seems better to prefer that mode when we have a choice.
- it avoids the dependency on `~`.
- it means you don't have to think about how the form of `y` isn't necessarily the same as the form of the argument (e.g. a `var`-form parameter accepts initializing expressions, but exposes a reference).

  F(~y);
}
```

then `SRC` is a value of `ConvertsInto(PF)` if and only if the call `F(~y)` is
legal.

Note that this depends on the meaning of the planned move operator `~`. In this
case, we need:

-   `~v` when `v` is a value parameter gives you the value of `v` before the
    expression;
-   `~r` when `r` is a `ref` parameter gives you a reference expression that
    references whatever `r` did before the expression;
-   `~i` allows you to forward the initializing expression used to initialize

>     `i` when `i` is a `var` parameter.

Reviewer: This phrasing makes it sound like e.g. if `PF` and `SRC` are `var T`, calling `F` will create only one `T` object, because the argument expression to `F` will somehow directly initialize the parameter of `G`. How about something like this instead?

```suggestion
-   `~i` when `i` is a `var` parameter gives you an initializing expression
    that initializes an object to the state that `i` was in before the expression.
```

}
```

This means `impl` lookup for queries involving `I` will consider `impl`
declarations that don't match exactly, unless the corresponding form values in
the `impl` declarations have
[a mark that requires them to match, such as `exact`](#modifiers-for-coherence).
In particular, in those positions a query `X` will match `Y` if `X` is a value
of type `ConvertsInto(Y)`. For example, the `impl` defined above acts as if it
was replaced by a collection of `impl`s for purposes of `impl` lookup:

```carbon
// `ForwardTo` is the `F` member of the `impl`
// above. It is equivalent to:
//   fn [ref self: u64](var a: i64) {
//     self += a;
//   }
alias ForwardTo =
  (form(ref u64) as I(form(var i64))).F;

impl forall [PF1:! ConvertTo(form(ref u64)),
             PF2:! ConvertTo(form(var i64))]
    PF1 as I(PF2) {
  fn F[self:? PF1](a:? PF2) {
    // This is allowed, by the definition of
    // `ConvertTo` and that the form of `self`
    // of `ForwardTo` is `ref _: u64` and
    // the form of its argument is `var _: i64`.
    (~self).(ForwardTo)(~a);
  }
}

impl forall [PF1:! ConvertTo(form(ref u64))]
   PF1 as I(exact form(var i64)) { ... }

impl forall [PF2:! ConvertTo(form(var i64))]

>    exact form(ref u64) as I(PF2) { ... }

Reviewer: ```suggestion
impl forall [PF1:! ConvertsInto(form(ref u64)),
             PF2:! ConvertsInto(form(var i64))]
    PF1 as I(PF2) {
  fn F[self:? PF1](a:? PF2) {
    // This is allowed, by the definition of
    // `ConvertsInto` and that the form of `self`
    // of `ForwardTo` is `ref _: u64` and
    // the form of its argument is `var _: i64`.
    (~self).(ForwardTo)(~a);
  }
}

impl forall [PF1:! ConvertsInto(form(ref u64))]
   PF1 as I(exact form(var i64)) { ... }

impl forall [PF2:! ConvertsInto(form(var i64))]
   exact form(ref u64) as I(PF2) { ... }
```

impl exact form(ref u64)
    as I(exact form(var i64)) {
  fn F[ref self: u64](var a: i64) = ForwardTo;
}
```

Note that these `impl` definitions are not created, but `impl` lookup acts as if
they exist. This replaces the `like` mechanism introduced in
[proposal #1144](https://github.com/carbon-language/carbon-lang/pull/1144), that
would add this parameterization to the `impl` itself, rather than making it part
of the `interface`.

Consequences:

-   The type structure of the match for purposes of prioritization counts
    `implicit_into` parameters as a pair of a type and the complete form, and
    changes depending on how well the query matches the `impl`.
    -   If the type of the `impl` doesn't match the type in the query, then both
        parts of the pair are considered to be `?`.
    -   If the type matches but rest of the form doesn't, the type structure is
        considered to be `(T, ?)`.
    -   If the whole form matches, ignoring `bound` which is not an aspect of
        forms that can be inferred, the type structure is considered to have a
        concrete match for both parts of the pair.

>     -   This allows a type to implement the same form interface twice, once with

Reviewer: Only twice? It can't implement for `var` separately as well?

        a value form and once with a reference form, and the one with the
        category matching the query will be chosen. (If the query is for an

>         initializing or constant expression, only the one with the value form

Reviewer: Why the value form? Converting an initializing form to a reference form requires fewer steps, so that seems like the more natural choice.

a parameter constraint:

```carbon
interface I[implicit_into Self:! Core.Form]
    (implicit_into P:! Core.Form) {
  fn F[self:? Self](a:? P);
}

class C(P:! I(form(val i16))) { ... }

fn G[P:! I(form(val i32))](c: C(P));
```

Unlike with interfaces with invariant parameters, multiple different `impl`s
could satisfy that constraint. In this case, the `C(P)` parameter type in `G`
would be legal with the existing constraint of `I(form(val i32))` on `P`, but
implied constraints would normally add another constraint that
`P impls I(form(val i16))`. When
[we considered this on 2025-05-29](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.dj0nqhfqo67w),
we decided we didn't want implied constraints adding constraints when there was
a choice of how to do it, so implied constraint don't apply to interfaces with
variant parameters.

#### Modifiers for coherence

Note that, unless modified by the `impl`, parameters marked `implicit_into`
can't be used to satisfy the orphan rule for coherence. As such we allow these
modifiers on those parameters, which can be used to ensure that there is a local
type in the `impl`'s type structure:

> -   `exact`: the type and category must match between the query and `impl`,

Reviewer: I understand why the type has to match, but why does the category have to match when the other aspects don't? That doesn't seem relevant to coherence or the orphan rule, and the reasoning for omitting the other aspects seems like it could apply to category as well.

  fn Op[self:? Self]
      (... each args:? each Param)
      ->? ResultForm;
}
```

Note that the `Self` is variant, since we want to allow a `val T` `impl` to be
used for a `ref T` object. The variance is using
[`anchor` modifier on `implicit_into`](#modifiers-for-coherence) so the _type_
of `Self` is not variant (though see
[the "Implicit conversions of `self`" section](#implicit-conversions-of-self)),
just the other aspects of the form.

#### `call`

It can be cumbersome to write out form values for the arguments to this
interface, though, and we expect this to be the most common use case that would
otherwise require writing form values. So we introduce a `call(`...`)`
convenience syntax for writing a facet type of the `Core.CallPrimitive`
interface.

| `call` syntax           | `Core.Form` facet type value                        |
| :---------------------- | :-------------------------------------------------- |
| `call()`                | `CallPrimitive()`                                   |
| `call(_: i32)`          | `CallPrimitive(form(val i32))`                      |
| `call(_: i32, _: bool)` | `CallPrimitive(form(val i32), form(val bool))`      |

> | `call(var _: bool)`     | `CallPrimitive(form(val bool))`                     |

Reviewer: I assume that in this context (unlike a parameter list) `var` functions as a modifier on a binding, not an independent pattern operator, so something like `call(var (_: bool, _: i32))` would not be valid? If so, it might be a good idea to explicitly say so.

| `call(bound _: i32*)`   | `CallPrimitive(form(bound val i32*))`               |
| `call(_:? X)`           | `CallPrimitive(X)`                                  |
| `call(... _: each T)`   | `CallPrimitive(... form(val each T))`               |
| `call() -> i32`         | `CallPrimitive() where .ResultForm = form(var i32)` |
| `call() -> val i32`     | `CallPrimitive() where .ResultForm = form(val i32)` |
| `call() -> ()`          | `CallPrimitive() where .ResultForm = form(var ())`  |

Note: If "`->` _\<return-type\>_" is not included at the end, the return type is
unconstrained. Use `-> ()` to mean "returns nothing."

Observe that `CallPrimitive` is `Core.Form` interface, so the above expressions

> represent `Core.Form` facet types. Alternatively you may specify the non-type

Reviewer: Is there any significance to calling these "`Core.Form` interfaces" rather than "form interfaces" as elsewhere in the doc?

> represent `Core.Form` facet types. Alternatively you may specify the non-type

Reviewer: The square brackets don't seem like an "alternative" to the above, they seem like a separate, composable aspect of the `call` feature. Maybe say "in addition" rather than "alternatively"?

a named form constraint, parallel to [form interfaces](#form-interfaces). This
overrides the default of `[Self:! type]`. A `require impls` or
`require Self impls` in a constraint is allowed when the type of `Self` matches
the type of the `Self` of the interface or constraint being required.

```carbon
constraint DefaultConstructibleCallback
    [Self:! Core.Form] {
  // `Self` is `Core.Form`, matches the `Self`
  // of `Callback` and `DefaultConstructible`.
  extend require impls Callback;
  extend require impls DefaultConstructible;
}
```

#### Associated constants in named constraints

We add support for associated constant declarations in named constraints, using
the same `let`...`:!`...`;` syntax as in interfaces. Such associated constants
must be used in such a way as to be deducible from the `impl`s of interfaces
required by the constraint. Deducible here means following the same rules as a
deduced parameter in a function signature. When seeing if a facet satisfies a
constraint, the associated constants are deduced.

```carbon
constraint Invalid {
  // ❌ Invalid: `Key` not deducible.
  let Key:! type;
  extend require impls Map(Key);
}

> ```

Reviewer: Could the definition of `MultiMap` be moved up here? I had a trouble following how the function signature rules would apply in this context, and that example would've helped a lot, but I didn't look that far down when I got confused.

and a desire to avoid a second syntax for expressing the prioritization between
`impl`s, we've decided to express this out-of-line in a named constraint using
the same `match_first` syntax used to prioritize `impl`s outside this context.

```carbon
constraint X[Self:! ParamForm] {
  match_first {
    require Self impls call(_: i32);
    require Self impls call(_: i64);
  }
}

fn J[T:! X](g:? T, x: i16) {
  // ✅ Uses `call(_: i32)`.
  g(x);
}
```

This was discussed on
[2025-05-29](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?pli=1&tab=t.0#heading=h.dj0nqhfqo67w).

#### `Core.Call` named constraint

We can use these new features to define `Core.Call` as a named constraint that
acts like the interface before this proposal, for use cases where the extra
control over forms is not needed.

```carbon
package Core;

constraint Call(... each Param:! type) {
  let Result:! type;
  extend require impls
      call[val](... _: each Param) -> Result;
  // Equivalent to:
  extend require impls
      ValSelf(
        CallPrimitive(... each form(val Param)))
      where .ResultForm = form(var Result);
}
```

> > FIXME: **QUESTION:** Is this redundant with providing `call`? Should instead

Reviewer: Does this question need to be resolved before this proposal is merged, or is it an open question for the future?

-   The "value `Self` returning a value" requirement can be satisfied due to the
    [`implicit_from` associated constant variance](#implicit_from-variance)
    allowing the reference return to be converted to a value return.
-   The "reference `Self` returning a reference" requirement takes advantage of
    [variance](#implicit_into-interface-parameter-variance) to support
    converting the reference `Self` to the value expected by the `impl`.

To support coherence, `IndexWithPrimitive` uses `implicit_into anchor` on the
`Self` parameter, since that is the parameter expected to be used to find the
`impl`. To support returning results that reference the `self` parameter, it is
marked `bound`.

See [Indexing](/docs/design/expressions/indexing.md) for details.

Switching the indexing interface to forms was
[discussed on 2025-06-10](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.stp5dpvithkz).

### Primitive member binding interface

Proposal
[#3720: Member binding operators](https://github.com/carbon-language/carbon-lang/pull/3720)
also introduced multiple interfaces in order to handle different expression

> categories.

Reviewer: The transition here is disorienting; I expected the following code to be an illustration of the interfaces introduced by 3720, not a redefinition of them.

```suggestion
categories. We can now redefine them as named constraints, in terms of a general `MemberBind` interface:
```

    `RightShiftWith`.
-   In
    [Comparison operators](/docs/design/expressions/comparison_operators.md#extensibility):
    `EqWith`, `OrderedWith`.

Note that the common type interfaces are unchanged.

### `auto` form

If `auto` is used in a position where a `Core.Form` is expected, such as in a
`:?` binding or `->?` return, the whole form is deduced, not just the type.

For `:? auto` parameters, specifically, the deduced form will not be `ref`
(using `val` instead) unless the argument expression is marked `ref`. See
[the "Deducing a form value" section](#deducing-a-form-value).

#### `auto` form bindings

This allows us to create a binding to another value without changing its phase
or category.

```carbon
let x:? auto = y;
// Equivalent to (if it was allowed):
let [P:! Core.Form] x:? P = y;

// In contrast,
let a: auto = b;
// only allows the type to vary
// and forces `a` to have value
// category and runtime phase.
```

This allows us to address the use case of creating a binding that is
[referentially transparent](https://en.wikipedia.org/wiki/Referential_transparency),
as considered in leads issues
[#996](https://github.com/carbon-language/carbon-lang/issues/996) and

> [#1371](https://github.com/carbon-language/carbon-lang/issues/1371). For

Reviewer: If I understand correctly, the binding itself isn't quite referentially transparent, because if `y` is an initializing expression, `x` will be a reference expression. Instead, `~x` is referentially transparent (given the assumptions you make elsewhere about the semantics of `~`). It's probably worth saying so explicitly.

contexts. This approach has limitations that we need to address, as described in
[leads issue #502: "More powerful aliasing feature"](https://github.com/carbon-language/carbon-lang/issues/5028).
The highest priority here is being able to give a name to a member of a
specific. For example, we've frequently assumed that declarations like
`alias String = BasicString(u8);`, `alias T = i32;` (equivalent to
`alias T = Core.Int(32);`), and `alias X = C(i32).X;` would somehow be allowed.
So we give `alias` with a definition in terms of `let` bindings instead of
operating on names:

> `alias A = B;` means `let A:? auto = B;`

An `alias` declaration is allowed in these contexts:

-   at file scope or in the body of a function;
-   in the body of a `class`, but only to define a compile-time constant;
-   in the body of an `impl` in order to specify that an associated function is
    to be implemented by an existing function; and
-   in the body of an interface or named constraint, where the rewrite to `let`
    is additionally given the `final` modifier.

An `alias` declaration is not allowed when the right-hand side (`B`) is a

> runtime expression, unless it is a reference expression with address known at

Reviewer: This should probably be inverted: the right-hand side must be a compile-time expression or a reference expression with address known at compile-time. As currently stated, this allows forming an alias to an expression with mixed phase (e.g. `(arr[v], BasicString(u8))` is not a runtime expression or a compile-time expression, so the current phrasing would allow it but mine would not).

  }
}

// Or: `FNT:! Core.Call() where .Result = i32`
fn ValueCaller[FNT:! call[val]() -> i32](f: FNT) -> i32 {
  return f();
}

Assert(ValueCaller(CallbackNoMutation) == 4);
Assert(ValueCaller(CallThroughPointer(&counter)) == 2);
```

NOTE: The `~` is there in anticipation of using moves to support forwarding, see
[the "Forwarding" section](#forwarding).

The first, generic approach is a bit more convenient but means using forms. The
second approach only uses types once `CallThroughPointer` has been defined, but
is less convenient. Some may prefer the second approach since it makes it
explicit when the callable object may be mutated.

### Forwarding

Forwarding is where an argument is passed along generically, including aspects
of its form like category and phase. Input-to-input forwarding is where the
arguments passed to one function become the arguments passed to another. For
example, this `Apply` function takes a callable object and argument expressions
to call it with and performs the call:

```carbon
fn Apply[... each ArgT:! Core.Form,
         FNT:! call(... _:? each ArgT)]
        (bound f:? FNT,
         bound ... each arg:? each ArgT)
        ->? FNT.ResultForm {

>   return (~f)(... ~each arg);

Reviewer: Why doesn't this need to constrain the types to be movable, the way the "future work" example below does?

    This proposal allows some programs and APIs to be written that couldn't
    otherwise be expressed in Carbon.
-   [Practical safety and testing mechanisms](/docs/project/goals.md#practical-safety-and-testing-mechanisms):
    The increased expressivity of forms allows us to prevent some bugs. For
    example, limiting conversion of `partial` types to initializing expressions
    avoids misuses of those values.

It does this while adhering to [Carbon's principles](/docs/project/principles):

-   [Principle: All APIs are library APIs](/docs/project/principles/library_apis_only.md):
    This machinery allows many more constructs to be expressed within the
    language as opposed to being special capabilities only available to
    built-ins. This includes abilities like how conversion to an adapter type
    can preserve category.
-   [Principle: Progressive disclosure](/docs/project/principles/progressive_disclosure.md):
    Developers don't need to be aware of this machinery unless they are engaged
    in code that needs these advanced features.

## Alternatives considered

> ### Decoupling type from the rest of the form

Reviewer: This seems like it belongs in #5545 (because it's about the basic structure of forms), and in particular it seems to overlap with the "Mixed expression categories" alternative discussed there. But on the other hand, here you're able to talk more concretely about how this choice affects form generics, whereas mine is phrased more speculatively because it notionally comes before yours. Maybe it's OK for my proposal to directly refer to yours, and so we should move this content to mine? Maybe we should continue to cover this alternative in both proposals, but cross-reference them somehow? What do you think?

that `in` and `out` were particularly expensive words to reserve, and which
corresponded to each kind of variance was hard to remember. The chosen keywords
are intended to emphasize what implicit conversion will be used.

We also considered making the variance keyword part of the parameter name, much
like `each` on variadic parameters. This seemed unnecessary verbosity given
that:

-   the compiler would enforce that the variance was consistent with usage,
    there was no choice about what variance to use in a given position; and
-   the `impl` you write is actual `impl` you get, there is just there's another
    layer around it.

We considered inferring variance, or just marking that there was variance and
inferring only the direction of variance. This had some downsides:

-   Variance comes with significantly different semantics, causing
    [facets to create adapters](#facets-create-adapters), so better if it is opt
    in.
-   Don't always want variance, for example since it requires extra annotations
    on impls to avoid coherence problems.
-   It goes against our

>     ["signature is the contract" Carbon principle](https://github.com/carbon-language/carbon-lang/pull/5990).

Reviewer: ```suggestion
    ["signature is the contract" (proposed) Carbon principle](https://github.com/carbon-language/carbon-lang/pull/5990). FIXME: update link when merged.
```

We also considered using no introducer inside `form(`...`)`, as in
`form(7 as i32)`, but it had some downsides:

-   Required lookahead when parsing.
-   We didn't want `form(i32)` to be legal and mean something other than
    `form(val i32)`, which is more likely what the code author intended.

Ultimately we didn't think that this was a common case that needed to be short.

We settled on [the `formof` approach](#formof). We expect that `formof(`...`)`
will be a valuable construct that we want outside of for compile-time value
forms.

For compound forms, we first tried to represent them using parens `(`...`)` and
curly braces `{`...`}` inside the `form(`...`)`. There was some concern about
ambiguity between the `()` value and the empty tuple form, similarly for `{}`,
but that was addressed by switching to `formof` for compile-time constant
values. However, the switch to `formof` meant that we needed a way to combine
`form(`...`)` and `formof(`...`)` expressions to create a compound form value.
The [compound form approach we settled on](#compound-form-values) had the
advantage of supporting all the composition we wanted without needing any

> dedicated syntax. It is a bit verbose, but that is perhaps acceptable given that

Reviewer: Why can't we instead say that `form()` can contain anything that can appear in a `->?` return form, e.g. `form((var i32, ref f64, formof(7)))`? That requires no more dedicated syntax than the current proposal, and it would make the language more consistent, as well as making compound form literals less verbose.

we considered `form(-> T)` instead of `form(var T)`, but we wanted to be
consistent about using keyword introducers for the different categories. We also
considered `form(init T)`, which was better for returns but worse for
parameters, but was not worth making `init` into a keyword.

In that discussion, we also considered `form(T)` instead of `form(val T)`, but
it seemed like it would lead to mistakes for returns where `val` was not the
default.

### `ref` on an argument controls its category

We considered an approach where `ref` on an argument would control whether a
reference expression was converted to a value expression before overload
resolution. Consider this example (using the overloading syntax proposed in
[discussion on 2025-03-28](https://docs.google.com/document/d/1Iut5f2TQBrtBNIduF4vJYOKfw7MbS8xH_J01_Q4e6Rk/edit?resourcekey=0-mc_vh5UzrzXfU4kO-3tOjA&tab=t.0#heading=h.t6l733mu79i7)):

```carbon
overload F {
  // Overload R
  fn (ref r: i32) -> i32;
  // Overload V
  fn (v: i32) -> bool;
}

var a: i32 = 1;
F(a);
```

With this approach, the call to `F` would convert the reference expression `a`
into a value before overload resolution and select "Overload V."

> We instead decided that `ref` on an argument only affects deduction. In this

Reviewer: ```suggestion
We instead decided that `ref` on an argument only affects deduction, because this
minimizes the divergence between function calls and other forms of pattern maching. In this
```

Feel free to take or leave that particular suggestion, but it would be good to give some indication of why we chose this.

    -   This new `alias` construct doesn't apply to namespaces, so a new
        namespace aliasing facility is introduced.

## Proposal

### Form type and values

First, we introduce the new special type `Core.Form` that describe the form of
bindings (such as function parameters or in a `let` declaration), returns, or
expressions.

-   The new `:?` syntax takes a form value to
    [specify the form of a binding](#specifying-the-form-of-a-binding).
-   The new `->?` syntax takes a form value to
    [specify the form of a return](#specifying-the-form-of-a-return).

Values of `Core.Form` are compile-time only, like types. Form values can be
compound forms or primitive forms. A compound form is either a tuple form or
struct form.

A primitive `Core.Form` has:

-   a type
-   a category, which can be `val`/value, `ref`/reference, or `var`/initializing
-   for parameter bindings, whether it is marked with `bound`
-   the phase, which can be "compile time" or "run time" (meaning "when the
    function is run")
-   for compile time phase, a specific compile-time value

#### Primitive form value literals

A primitive `Core.Form` value with "run time" phase can be written `form(`_cat_

> _type_`)`, where _cat_ is an optional category modifier, and _type_ is a type

Reviewer: If it's optional, we should explicitly specify the default, but I don't think it should be optional. Even if `form(i32)` is valid, an implicit default costs the reader much more than it gains the writer.

(I continue to think it would really be nice to write composite form literals like `form((ref T, ref U))` rather than `(form(ref T), form(ref U))`, but that's different from having a default category for primitive forms.)

### Specifying the form of a binding

The `n :? B` binding syntax creates a binding to the name `n` with form given by
`B`. Note that `B` will be implicitly converted to `Core.Form`, much like a `:`
binding will implicitly convert its right-hand argument to `type`. Example:

```carbon
fn VarFn(x:? form(var i32));
// is equivalent to:
fn VarFn(var x: i32);

fn RefFn(x:? form(ref i32));
// is equivalent to:
fn RefFn(ref x: i32);
```

While this is valid, this is not how this feature is expected to be used.
Instead, the form value will typically be a generic parameter.

```carbon
fn Explicit(PF:! Core.Form, arg:? PF);

//  ✅ Valid since `5` can be passed to a
// `var _: i32` parameter.
Explicit(form(var i32), 5);

// ❌ Invalid, `5` is not a reference
// expression.
Explicit(form(ref i32), 5);

// ❌ Invalid, `true` can't be cast to `f32`.
Explicit(form(val f32), true);
```

A `Core.Form` parameter is deducible from an argument expression, see
[the "Deducing a form value" section](#deducing-a-form-value). In fact, the form
of a binding is _about_ the expressions it binds to. In contrast, the result of
a `var` and `ref` binding is a reference expression in both cases.

> A `:?` binding with a keyword modifier, like `ref` or `bound`, forces that

Reviewer: This should also address nesting `:?` within other pattern operators, most notably `var` (but note that the current guidance on #6342 is to treat `ref` as a pattern operator as well).

I'd recommend saying that modifiers on `:?` are disallowed by default, with `bound` (or its successor) called out as an exception. Similarly, I'd recommend saying that a `:?` pattern can't be directly or indirectly nested inside a `var` pattern (or `ref` pattern, if `ref` is a pattern operator). There are multiple plausible interpretations for patterns like `ref x:? form(val i32)` or `var x:? form(val i32)`, so unless we have clear use cases, it seems safer to disallow them. It also seems surprising to have syntaxes for coercing a form binding to template phase, or reference or initializing category, but no corresponding syntaxes for coercing it to symbolic or runtime phase, or value expression category.

See also discord discussion [here](https://discord.com/channels/655572317891461132/748959784815951963/1473464233697218611)

```carbon
interface Callback[Self:! Core.Form] {
  // Uses `Self` as the form of a
  // parameter using `:?`.
  fn Go[self:? Self]();
}

interface DefaultConstructible
    [Self:! Core.Form] {
  // Uses `Self` as the form of the
  // return using `->?`.
  fn MakeDefault() ->? Self;
}
```

`Callback` and `DefaultConstructible` are _form interfaces_, specifically
`Core.Form` interfaces in this case. A value of a form interface is a _form
facet_. These work the same way as ordinary "type" interfaces except they use
the matching kind of form values instead of type values. An ordinary "type"
interface acts as if `Self` is declared `[Self:! type]` instead of with a form
type.

Let's break down what happens with an ordinary "type" interface:

```carbon
// Equivalent to: interface I[Self:! type]
interface I {
  fn F[self: Self]();
}

// Since `I` is effectively declared `I[Self:! type]`, that means
// `T` is a facet of a type, which means a type that is known to
// implement `I`. Since `T` is a facet of a type, it has an implicit
// conversion to `type`, so it can be used to the right of a `:`
// to give a type to `x`.
fn A[T:! I](x: T) {

>   x.F():

Reviewer 2: Should this be a semicolon here instead of a colon?
